consensus/core/src/tests/randomized_tests.rs

```diff
@@ -0,0 +1,233 @@
+/// on all waves as we may have an Undecided leader somewhere early in the sequence.
+#[test]
+fn test_randomized_dag_and_decision_sequence() {
+    let mut random_test_setup = random_test_setup();
+
+    // Setup for Authority 1
+    let mut authority_1 = authority_setup(random_test_setup.num_authorities);
+
+    let pipeline = random_test_setup.num_waves % DEFAULT_WAVE_LENGTH as usize;
+    let wave_number =
+        authority_1.committer.committers[pipeline].wave_number(random_test_setup.num_waves as u32);
+    let num_rounds = authority_1.committer.committers[pipeline].decision_round(wave_number);
+    let include_leader_percentage = 50;
+    let dag_builder = create_random_dag(
+        random_test_setup.seed,
+        include_leader_percentage,
+        num_rounds,
+        authority_1.context.clone(),
+    );
+
+    tracing::info!(
+        "Running test with committee size {} & {} completed waves in the DAG...",
+        random_test_setup.num_authorities,
+        random_test_setup.num_waves
+    );
+
+    let mut all_blocks = dag_builder.blocks.values().cloned().collect::<Vec<_>>();
+    all_blocks.shuffle(&mut random_test_setup.seeded_rng);
+
+    let mut sequenced_leaders_1 = vec![];
+    let mut last_decided = Slot::new_for_test(0, 0);
+    for block in &all_blocks {
+        let _ = authority_1
+            .block_manager
+            .try_accept_blocks(vec![block.clone()]);
+        let sequence = authority_1.committer.try_decide(last_decided);
+
+        if !sequence.is_empty() {
+            sequenced_leaders_1.extend(sequence.clone());
+            let leader_status = sequence.last().unwrap();
+            last_decided = Slot::new(leader_status.round(), leader_status.authority());
+        }
+    }
+
+    assert!(authority_1.block_manager.is_empty());
+
+    // Setup for Authority 2
+    let mut authority_2 = authority_setup(random_test_setup.num_authorities);
+
+    let mut all_blocks = dag_builder.blocks.values().cloned().collect::<Vec<_>>();
+    all_blocks.shuffle(&mut random_test_setup.seeded_rng);
+
+    let mut sequenced_leaders_2 = vec![];
+    let mut last_decided = Slot::new_for_test(0, 0);
+    for block in &all_blocks {
+        let _ = authority_2
+            .block_manager
+            .try_accept_blocks(vec![block.clone()]);
+        let sequence = authority_2.committer.try_decide(last_decided);
+
+        if !sequence.is_empty() {
+            sequenced_leaders_2.extend(sequence.clone());
+            let leader_status = sequence.last().unwrap();
+            last_decided = Slot::new(leader_status.round(), leader_status.authority());
+        }
+    }
+
+    assert!(authority_2.block_manager.is_empty());
+
+    // Ensure despite the difference in when blocks were received eventually after
+    // receiving all blocks both authorities should return the same sequence of blocks.
+    assert_eq!(sequenced_leaders_1, sequenced_leaders_2);
+}
```

Comment:
Very nice tests! I would suggest a couple of things (feel free to ignore them for now):
1. Do many iterations for each test - we currently test one random DAG. I think it will be better if we try to produce a few hundreds to test. Something that would make sense for running as part of a unit test suite (so doesn't take too much time). Using a seed chain would probably do for reproducibility.
2. Randomize the number of blocks fed to block manager. We currently provide those one by one. Maybe we could do this randomly as well and pass `N` number of blocks per time